apps/prairielearn/elements/pl-file-upload/pl-file-upload.js

```diff
@@ -11,20 +20,57 @@
 
   class PLFileUpload {
     constructor(uuid, options) {
+      // Not configurable at this point as this matches the request size limit enforced by the server
+      this.maxFileSizeMB = 5;
       this.uuid = uuid;
       this.files = [];
-      this.acceptedFiles = options.acceptedFiles || [];
-      this.acceptedFilesLowerCase = this.acceptedFiles.map((f) => f.toLowerCase());
+      this.requiredFiles = options.requiredFiles || [];
+      this.requiredFilesLowerCase = this.requiredFiles.map((f) => f.toLowerCase());
+      this.requiredFilesRegex = options.requiredFilesRegex || [];
+      // Initialized after files are downloaded
+      this.requiredFilesUnmatchedRegex = this.requiredFilesRegex.slice();
+      this.optionalFiles = options.optionalFiles || [];
+      this.optionalFilesLowerCase = this.optionalFiles.map((f) => f.toLowerCase());
+      this.optionalFilesRegex = options.optionalFilesRegex || [];
+
+      // Checks whether a file name is acceptable
+      // If yes, it returns the canonical name of the file, if not, it returns null
+      // Note the priority order: first, fill required file names, then required patterns, then optional names, then optional patterns
+      this.findAcceptedFileName = (fileName) => {
+        if (this.requiredFilesLowerCase.includes(fileName)) {
+          return this.requiredFiles[this.requiredFilesLowerCase.indexOf(fileName)];
+        }
+        if (this.requiredFilesUnmatchedRegex.some((f) => new RegExp(f[0], 'i').test(fileName))) {
+          return fileName;
+        }
+        if (this.optionalFilesLowerCase.includes(fileName)) {
+          return this.optionalFiles[this.optionalFilesLowerCase.indexOf(fileName)];
+        }
+        if (this.optionalFilesRegex.some((f) => new RegExp(f[0], 'i').test(fileName))) {
+          return fileName;
+        }
+        return null;
+      };
+
       this.pendingFileDownloads = new Set();
       this.failedFileDownloads = new Set();
 
       const elementId = '#file-upload-' + uuid;
       this.element = $(elementId);
       if (!this.element) {
-        throw new Error('File upload element ' + elementId + ' was not found!');
+        throw new Error(`File upload element ${elementId} was not found!`);
       }
 
       if (options.submittedFileNames) {
+        options.submittedFileNames.forEach((n) => {
+          if (this.requiredFiles.includes(n)) return;
+          const matchingRegex = this.requiredFilesUnmatchedRegex.findIndex(
+            (f) => new RegExp(f[0], 'i').test(n) && this.requiredFilesUnmatchedRegex.includes(f),
```

Comment:
Why is the includes check necessary? The callback will only be called with entries in the array.

Reply:
It is not necessary. I think this is leftover from some earlier version where this check was more complicated.